pricegraph/src/orderbook.rs

```diff
@@ -185,6 +219,43 @@ fn as_effective_sell_price(price: &Price) -> f64 {
     FEE_FACTOR * (price.denominator as f64) / (price.numerator as f64)
 }
 
+/// A type definiton for a mapping between user IDs to user data.
+type UserMap = HashMap<UserId, User>;
+
+/// User data containing balances and number of orders.
+#[derive(Debug, Default, PartialEq)]
+struct User {
+    /// User balances per token.
+    balances: HashMap<TokenId, f64>,
+    /// The number of orders this user has.
+    num_orders: usize,
+}
+
+impl User {
+    /// Adds an encoded orderbook element to the user data, including the
+    /// reported balance and incrementing the number of orders.
+    fn include_order(&mut self, element: &Element) -> usize {
+        let order_id = self.num_orders;
+
+        self.balances
+            .entry(element.pair.sell)
+            .or_insert_with(|| u256_to_f64(element.balance));
```

Comment:
Is it worth asserting here that if the element already exists it's equal to `element.balance`?

Reply:
It would validate the data a bit which is nice.

Reply:
So, I looked into bubbling up that error and its fairly disruptive. I'll add it as a separate PR as it would affect the code a bit.

In that case we should also validate other order parameters like making sure that the orders are actually valid and that the the price denominator is non-zero, etc.

Reply:
Created #710

Reply:
Fine to do later. We might even be able to panic instead of bubbling up some error, as this is really just a sign of inconsistent data.